### PR TITLE
[TTIR, TTIR -> TTNN] Fix slice verification

### PR DESCRIPTION
**Pull request by PavleSarenac**
### Ticket
[**Issue 6275**](https://github.com/tenstorrent/tt-mlir/issues/6275): Fix slice verification

### Problem description
Before this pull request, TTIR was not fully compatible with PyTorch slicing semantics. PyTorch slicing never raises errors for out-of-bounds slice indices, so TTIR should behave the same way.

Since this pull request relaxes verification for the TTIR static slice operation, an appropriate conversion pattern is required before lowering to the TTNN dialect to ensure slice bounds are handled correctly and to prevent out-of-bounds errors.

Below is the analysis of the compatibility between PyTorch and TTIR slicing before this pull request:

before_the_fix

Below is the analysis of the compatibility between PyTorch and TTIR slicing after this pull request:

after_the_fix_pytorch_vs_ttir

### What's changed

- `tt-mlir/lib/Dialect/TTIR/IR/TTIROps.cpp`: relaxed `mlir::tt::ttir::SliceStaticOp` verification to match PyTorch behavior in TTIR.
- `tt-mlir/lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp`: implemented `SliceStaticOpConversionPattern`to compensate for relaxed `mlir::tt::ttir::SliceStaticOp` verification to ensure that valid slicing is delivered to TTNN as before.
- `tt-mlir/test/python/golden/ttir_ops/data_movement/test_data_movement.py`: added comprehensive tests to validate all the edge cases regarding the new behavior in the slicing compilation pipeline.
- `tt-mlir/test/ttmlir/Dialect/TTIR/data_movement/slice/slice_tests_negative.mlir`: removed tests that have become invalid and incorrect after this pull request.
- `tt-mlir/test/ttmlir/Dialect/TTIR/data_movement/slice/slice_tests_positive.mlir`: added comprehensive tests to validate all the edge cases regarding the new behavior in the slicing compilation pipeline.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: test/python/golden/ttir_ops/data_movement/test_data_movement.py [hunk 1]
```diff
@@ -387,7 +387,18 @@ def hoisted_reshape_wrapper(
...
+        ((5,), [-10], [-8], [1]),
+        ((5,), [-10], [-12], [1]),
+        ((5,), [-5], [-2], [1]),
+        ((5,), [-15], [15], [1]),
+        ((5,), [2], [10], [1]),
+        ((5,), [10], [12], [1]),
+        ((5,), [12], [10], [1]),
+        ((5,), [5], [5], [1]),
         # Strided 1D
         ((70,), [3], [62], [7]),
```
[the remark this answers is not in the document]
Comment (PavleSarenac) [reply test/python/golden/ttir_ops/data_movement/test_data_movement.py:403]: This has been an already existing limitation before this pull request. If a negative step is added to these tests, `builder.base.builder_runtime.TTBuilderCompileException` exception is raised with "step must be greater than zero" message.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 2]
```diff
@@ -909,19 +909,96 @@ class ReshapeOpConversionPattern : public OpConversionPattern<ttir::ReshapeOp> {
...
+      int32_t begin = ::mlir::cast<::mlir::IntegerAttr>(begins[dim]).getInt();
+      int32_t end = ::mlir::cast<::mlir::IntegerAttr>(ends[dim]).getInt();
+      int32_t step = ::mlir::cast<::mlir::IntegerAttr>(stepAttr[dim]).getInt();
+
+      // Adjust begin and end for a positive step
+      int32_t adjustedBeginPositiveStep = (begin < kDimStart) ? 
+        std::max<int32_t>(begin + dimSize, kDimStart) : 
+        std::min<int32_t>(begin, dimSize);
+      int32_t adjustedEndPositiveStep = (end < kDimStart) ? 
+        std::max<int32_t>(end + dimSize, kDimStart) :
```
[the remark this answers is not in the document]
Comment (PavleSarenac) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: I have included it directly now.

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 3]
```diff
@@ -3140,105 +3140,75 @@ ::mlir::LogicalResult mlir::tt::ttir::SliceStaticOp::verify() {
...
+      std::min<int32_t>(end, dimSize);
+    if (adjustedBeginPositiveStep > adjustedEndPositiveStep) {
+      // e.g. array[3:1:1] <=> array[0:0:1]
+      adjustedBeginPositiveStep = adjustedEndPositiveStep = kDimStart;
+    }
+    
+    // Adjust begin and end for a negative step
+    int32_t adjustedBeginNegativeStep = (begin < kDimStart) ? 
+      std::max<int32_t>(begin + dimSize, kDimStart) : 
+      std::min<int32_t>(begin, dimSize - 1);
```
[the remark this answers is not in the document]
Comment (PavleSarenac) [reply lib/Dialect/TTIR/IR/TTIROps.cpp]: I agree, fixed it.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 4]
```diff
@@ -909,19 +909,96 @@ class ReshapeOpConversionPattern : public OpConversionPattern<ttir::ReshapeOp> {
...
+        std::min<int32_t>(end, dimSize);
+      if (adjustedBeginPositiveStep > adjustedEndPositiveStep) {
+        // e.g. array[3:1:1] <=> array[0:0:1]
+        adjustedBeginPositiveStep = adjustedEndPositiveStep = kDimStart;
+      }
+      
+      // Adjust begin and end for a negative step
+      int32_t adjustedBeginNegativeStep = (begin < kDimStart) ? 
+        std::max<int32_t>(begin + dimSize, kDimStart) : 
+        std::min<int32_t>(begin, dimSize - 1);
```
[the remark this answers is not in the document]
Comment (PavleSarenac) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp]: I agree, fixed it.

**Comment by PavleSarenac** [comment]
Hey @mvasiljevicTT I have pushed a [**new commit**](https://github.com/tenstorrent/tt-mlir/pull/9026/changes/aaf07ae5da8a806374bcd3e12d519edcd2690a43) that should fix [**the error**](https://github.com/tenstorrent/tt-mlir/actions/runs/29591533411/job/87926672535) related to code formatting from the CI.

Also, I saw that there was [**one more error**](https://github.com/tenstorrent/tt-mlir/actions/runs/29591533411/job/87926785905) in the CI pipeline, but this seems to be unrelated to my changes.

Could you please run the CI pipeline again?

**Comment by PavleSarenac** [comment]
Hi @mvasiljevicTT I am writing to follow up on this PR. 

Since I have seen that all CI checks are passing, I would like to know if you have maybe had a chance to review the code changes and whether there are any modifications needed on my end in order to have this PR merged. 

Thanks in advance for the feedback.
